### PR TITLE
fix: do not create metrics collector in each loop

### PR DESCRIPTION
**Pull request by youtuyy**
**Issue #, if available:**
Fix the bug reported in #3 

**Description of changes:**
Create a new metrics_collector only when the workflow initialized

**Why is this change necessary:**
Before, the metrics_collector was initialized once in each loop. In this way, the [delta metrics](https://github.com/aws-samples/aws-iot-device-defender-agent-sdk-python/blob/a3d27c51311e12625bd0b774e8b730ead3bc515d/AWSIoTDeviceDefenderAgentSDK/metrics.py#L31-L32) (e.g. bytes in or bytes out) could not be calculated. So we only initialize a new object of metrics_collector at initialization. 

**How was this change tested:**
Created a test component with the code and deployed. Checked the delta was correctly published.

**Any additional information or context required to review the change:**

**Checklist:**
 - [ ] Updated the README if applicable
 - [ ] Updated or added new unit tests
 - [ ] Updated or added new integration tests
 - [ ] Updated or added new end-to-end tests
 - [ ] If your code makes a remote network call, it was tested with a proxy
 - [ ] You confirm that the change is backwards compatible

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: greengrass_defender_agent/agent.py [hunk 1]
```diff
@@ -67,33 +67,43 @@ def wait_for_config_changes(ipc_client):
...
 
     :param ipc_client: Ipc client
     :param new_config: a configuration object
+    """
+    config.logger.info("Collector running on device: {}".format(config.THING_NAME))
+    config.logger.info("Metrics topic: {}".format(config.TOPIC))
+    sample_interval_seconds = new_config[config.SAMPLE_INTERVAL_NEW_CONFIG_KEY]
+    config.logger.info("Sampling interval: {} seconds".format(sample_interval_seconds))
+
+    metrics_collector = collector.Collector(short_metrics_names=False)
```
Comment (MikeDombo) [inline greengrass_defender_agent/agent.py:82]: Even if config changed, none of the config affects this object so don't create it again. Create it only 1 time in the constrictor of a class

Comment (youtuyy) [reply greengrass_defender_agent/agent.py]: Updated. 
I had weighed both options. Was thinking the user might want everything "refreshed" when changing the configuration. And the first delta after a reconfiguration would not represent the delta for a full interval. But after reconsidering I agree it makes sense to only have one collector object. The system metrics' delta should preserve despite configuration change.

**Review by randolphtan121 — COMMENTED** [review]
Code changes here look good, how are we validating that the confirmed changes are now calculating the correct delta?

**Review by MikeDombo — APPROVED** [review]
All of this should be refactored into a proper class instead of loose functions so that you don't need to pass around all of those arguments. They really ought to be instance members.

Approving for now as we need the fix, but refactoring should follow.

**Review by randolphtan121 — APPROVED** [review]
Approving given the fact that validation is being done manually.

**Comment by youtuyy** [comment]
Agreed to Michael. Should refactor the main function into a proper class. Will track the refactoring task in back log. Merge to get the fix for now

**Comment by olivierpichon** [comment]
Hey guys, was wondering if you could please provide an update on this. This PR was merged and it looks like the documentation was updated to reflect on the new version: https://docs.aws.amazon.com/greengrass/v2/developerguide/device-defender-component.html#device-defender-component-changelog
So I started to try using it, and got the following error:
```
java.util.concurrent.ExecutionException: com.aws.greengrass.componentmanager.exceptions.NoAvailableComponentVersionException: No local or cloud component version satisfies the requirements. Check whether the version constraints conflict and that the component exists in your AWS account with a version that matches the version constraints. If the version constraints conflict, revise deployments to resolve the conflict. Component aws.greengrass.DeviceDefender version constraints: thing/olis-gateway requires =3.0.1
```

Thanks for your help!

**Comment by MikeDombo** [comment]
Yes this change has been merged, however due to re:Invent, which is ongoing, we are not releasing the change yet. 
It should be released next Monday, but no guarantee for that date.

If you wish, you may take the code from this repository and make your own copy of the component in your account in order to get up and running again.

**Comment by olivierpichon** [comment]
Thanks @MikeDombo !

**Pull request merged**